src/actions/groups.cr

```diff
@@ -1,34 +1,150 @@
+      %[Group ID #{group_id_s} was not a valid integer]
+    # find the group if it already exists in the permissions config
+    index = DppmRestApi.permissions_config
+      .groups
+      .index { |grp| grp.id == group_id }
+    raise NoSuchGroup.new context, group_id_s if index.nil?
+    old_group = DppmRestApi.permissions_config.groups[index]
+    path = URI.unescape context.params.url["path"]
+    access_level = Access.parse?(context.params.url["access_level"])
+    raise InvalidAccessLevel.new context if access_level.nil?
+    name = context.params
+      .query["name"]?
+      .try { |param| URI.unescape param } || old_group.name
+    # Create the group if it doesn't already exist
+    group = Config::Group.new id: group_id, name: name,
+      permissions: old_group.permissions
+    # Create a query parameters object (since the default isn't nil, it's an
+    # empty Hash) if group doesn't already have a path with query params on
+    # it.
+    existing_params = group.permissions[path]?
+      .try &.query_parameters || {} of String => Array(String)
+    # if there were query parameters specified in the body, override
+    # existing values
+    existing_params.merge! query_params_in_body?(of: context) do |_path, existing, new_val|
+      existing + new_val
+    end
+    # Override the existing permissions
+    group.permissions[path] = Config::Route.new access_level, existing_params
+    # and overwrite the current value.
+    DppmRestApi.permissions_config.groups[index] = group
   end
-  relative_put "/:id/param" do |context|
+  # Remove specific query parameters from this path/access-level for this
+  # group. The query parameters may be specified via query parameters like:
+  # ```
+  # ?some-param=value&some-param=value2
+  # ```
+  # Not specifying any query parameters will delete all parameter
+  # specifications on this route.
+  #
+  # Not specifying a list of values...
+  # ```
+  # ?some-param&some-other-param
+  # ```
+  # ...will delete that parameter from the group.
+  #
+  # So, in summary the request...
+  # ```
+  # /123/param/%2Fsome%2Fpath?some-param=value&param2&param3&param4=value2&param4=value3
+  # ```
+  # ...(if succesful) would result in the deletion of the following query
+  # parameter permissions on `/some/path` for group `123`:
+  #  - access to "value" on the query parameter  "some-param"
+  #  - access to any route which requires `param2` or `param3` to be specified
+  #  - access to "value2" and "value3" on "param4"
+  relative_delete "/:id/param/:path" do |context|
     raise Unauthorized.new context unless Actions.has_access? context, Access::Update
-    # TODO: add a new param to the group
-
+    path = context.params.url["path"]
+    group_id_s = context.params.url["id"]
+    group_id = group_id_s.to_i? || raise BadRequest.new context,
+      %[Group ID "#{group_id_s}" was not an integer!]
+    group_idx = DppmRestApi.permissions_config.groups.index { |grp| grp.id == group_id }
```

Comment:
`reject!` can be used, then raise if the array is still the same.
This will also ensure that no other remaining group still have the ID.

Reply:
This route isn't deleting the group, it's deleting access to this path for this group. `#reject!` wouldn't work... I could see `#map` like
```crystal
group_id_already_found = false
DppmRestApi.permissions_config.groups.map do |group|
  next group if group.id != group_id
  raise InternalServerError.new context, "duplicate group with id " + group_Id_s if group_id_already_found
  group_id_already_found = true
  group.remove_access_to_path context.params.path["path"]
  group
end
DppmRestApi.permissions_config.sync_to_disk
```
... or using `#update` like `DppmRestApi.permissions_config.groups.update group_idx do |group|...`

Reply:
Right, sorry